contracts/EscrowFixedPrice.sol

```diff
@@ -66,65 +76,121 @@ contract EscrowFixedPrice is EscrowBase {
+        }
+
+        // Adjust account balance
+        balance -= maxDeliveryRate;
+        mapAccountBalances[msg.sender][token] = balance;
+    }
+
+    // TODO buyBackBurner does not account for other tokens but WETH, OLAS
+    /// @dev Drains collected fees by sending them to a Buy back burner contract.
+    function drain(address token) external virtual override {
+        // Reentrancy guard
+        if (_locked > 1) {
+            revert ReentrancyGuard();
+        }
+        _locked = 2;
+
+        uint256 localCollectedFees = mapCollectedFees[token];
+
+        // Check for zero value
+        if (localCollectedFees == 0) {
+            revert ZeroValue();
+        }
+
+        mapCollectedFees[token] = 0;
+
+        // Check token address
+        if (token == address (0)) {
+            // Wrap native tokens
+            _wrap(localCollectedFees);
+            // Transfer to Buy back burner
+            IToken(wrappedNativeToken).transfer(buyBackBurner, localCollectedFees);
+        } else {
+            IToken(token).transfer(buyBackBurner, localCollectedFees);
+        }
+
+        emit Drained(token, localCollectedFees);
+
+        _locked = 1;
+    }
+
+    function _withdraw(address token, uint256 balance) internal {
+        bool success;
+        // Transfer mech balance
+        if (token == address(0)) {
+            // solhint-disable-next-line avoid-low-level-calls
+            (success, ) = msg.sender.call{value: balance}("");
+        } else {
+            IToken(token).transfer(msg.sender, balance);
+        }
+
+        // Check transfer
+        if (!success) {
+            revert TransferFailed(token, address(this), msg.sender, balance);
         }
     }
 
     /// @dev Withdraws funds for a specific mech.
-    function withdraw() external virtual override {
+    function withdrawMech(address token) external virtual override {
         // Reentrancy guard
         if (_locked > 1) {
             revert ReentrancyGuard();
         }
         _locked = 2;
 
         // Get mech balance
-        uint256 balance = mapMechBalances[msg.sender];
+        uint256 balance = mapMechBalances[msg.sender][token];
+        // TODO limits?
         if (balance == 0) {
             revert UnauthorizedAccount(msg.sender);
         }
 
-        // Transfer mech balance
-        // solhint-disable-next-line avoid-low-level-calls
-        (bool success, ) = msg.sender.call{value: balance}("");
-        if (!success) {
-            revert TransferFailed(address(0), address(this), msg.sender, balance);
-        }
+        _withdraw(token, balance);
 
-        emit Withdraw(msg.sender, balance);
+        emit Withdraw(msg.sender, token, balance);
 
         _locked = 1;
     }
 
-    /// @dev Drains collected fees by sending them to a Buy back burner contract.
-    function drain() external virtual override {
+    /// @dev Withdraws funds for a specific account.
+    function withdrawAccount(address token) external virtual override {
```

Comment:
So this function allows to withdraw from a token and the limit of amount of this withdrawal is the balance of the requester in the marketplace (so if all is used for requests nothing can be withdrawn using this function). But how is it ensured that the requester can't withdraw from this token by other means than through this the marketplace and this function ?

Reply:
This is outdated now :)

Reply:
Balances are checked - you can't withdraw somehow from another place.